### ISSUE TITLE
[feature] Introduce default components

### ISSUE THREAD

**Issue by lmglmg** [conan-io/conan#14751]
### What is your suggestion?

Currently, Conan packages can have various components, and users can cherry-pick which ones they want to use. However, if a consumer package doesn't explicitly specify which components it needs, it ends up including all of them. This can lead to issues, especially when some components are hefty or are meant only for testing.

Now, here's my idea: Let's introduce the concept of default components in Conan, somewhat akin to how Rust's default features work. In Rust, crates (packages) have features (components), and each crate can specify default features. When users import a crate, they automatically get the default features along with any features they manually select. And if users want to skip the default features, they can simply specify `default-features = false` in their `Cargo.toml` file.

I believe we can adopt a similar mechanism in Conan, where we define default components for packages. Here's how it could work:

1. Defining Default Components: Package creators can specify which components should be considered default for their packages. This information would be stored in the package recipe.

2. User-Friendly Defaults: When users pull in a Conan package, they'd automatically get the default components, making it easier for them to get started without specifying every component individually.

3. Manual Component Selection: Users can still manually select the components they need, just like they do now. This won't change.

4. Opting Out of Defaults: For backward compatibility, if no default components are specified in the package recipe, we can consider all components as default components. This ensures that existing packages still work seamlessly.

This approach simplifies the Conan package usage for common scenarios, making it more user-friendly while retaining the flexibility for advanced use cases.

Rust features documentation : https://doc.rust-lang.org/cargo/reference/features.html

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @lmglmg 

Thanks for your suggestion.
I think this can be a valid feature request, and it would make sense for some scenarios. My main concern would be the added complexity vs added value. The ``components`` processing at this time is quite complicated, and while we have wanted to try to refactor, simplify and improve it a bit, we haven't managed to do so yet. So this feature would need to be built on top of that, making the effort (specially the later derived maintenance and support) relatively large, compared to the number of cases that it would be addressing.

It is important that there is built-in approach to this: packages consuming this "components" dependency package, can explicitly define which components they use. While this could be a bit repetitive, it is a viable and relatively simple approach. So all in all, this wouldn't be a high priority, I am tagging it for the 2.X roadmap to also give time for extra users feedback.

**Comment by DoDoENT** [comment]
Hi @memsharded , 

I don't think this needs to be a complicated feature. It can be akin to `package_type` in that it will only serve for the default initialization of the `requires` attribute of the `cpp_info` object in the `package_info` method.

Here is the example that we have:

- package `protection` contains two components: `protection` and `keygen`. The `protection` component contains algorithms for verifying the license key, and the `keygen` component contains algorithms for generating license keys. To avoid accidental use of the `keygen` component from the `protection` package, I added `self.cpp_info.set_property('cmake_target_name', 'protection::all')`, however, this only affects cmake, not Conan, as I'll explain below
- there is a package `feature` that contains no components. It depends on the `protection` to get license verification service and also uses `gtest` for its unit tests. For purposes of unit tests, a temporary license key needs to be generated, so the unit test target within the cmake links with `GTest::gtest` and `protection::keygen`. However, it's very easy to write a `package_info` of the `feature` package like this:

```python
def package_info(self):
    self.cpp_info.libs = ['feature']
```

Now, this contains a serious bug, because a user of the `feature` package gets both `protection::protection` and `protection::keygen` in its dependency graph. Thus, if this user is some final app that gets shipped to the customers, it contains both algorithms for generating the license key (keygen) and license key verification. Cracking such an app is then very easy.

Unfortunately, the app has no way of preventing `protection::keygen` from getting linked into the binary. The culprit is the `feature` package, as defined above. The correct implementation of its `package_info` method should look like this:

```python
def package_info(self):
    self.cpp_info.libs = ['feature']
    self.cpp_info.requires = ['protection::protection']
```

As you can see, this is very cumbersome and requires a lot of vigilance.

The suggestion by @lmglmg would allow for the `protection` package to define a set of default components that would then be used to automatically initialize the `requires` attribute of the `cpp_info` object in the `feature` package's `package_info` method, so a simple `package_info` could be used.

For example, imagine if `protection` package could write something like this in it's `package_info` method:

```python
def package_info(self):
    self.cpp_info.default_components = ['protection::protection']

    # setup self.cpp_info.components['protection'] 
    # setup self.cpp_info.components['keygen']
```

Then, `feature` could be written as simply as 

```python
def package_info(self):
    self.cpp_info.libs = ['feature']
```

and that would have the same meaning as

```python
def package_info(self):
    self.cpp_info.libs = ['feature']
    self.cpp_info.requires = ['protection::protection']
```

Of course, if `feature` sets custom values to its `self.cpp_info.requires`, this will take precedence. The idea is just to have default-initialization of the `requires` attribute to avoid easy-to-make mistakes as the one I described above.

What do you think about such a proposal?

**Comment by memsharded** [comment]
Hi @lmglmg 

The ``default_components`` have been introduced in Conan 2.9 and Conan 2.10 and are used by the new ``CMakeDeps`` generator.
Please check this new feature in https://docs.conan.io/2/incubating.html. It is in dev-testing stage, it means intended for initial testing by users, so feedback very welcome.

I am closing this ticket as implemented in Conan 2.10, please open new tickets for any further question, issue or feedback you might have. Thanks very much for your feedback!
